### ISSUE TITLE
Build C++ custom module as shared library failed on macOS

### ISSUE THREAD

**Issue by haxpor** [godotengine/godot-docs#913]
Related to https://github.com/godotengine/godot-docs/issues/499 but for macOS (seems Linux is fine to compile shared library).

Note, this is based on 2.1-stable source code.
I successfully compile C++ custom module as static library on macOS 10.13 (17A365) but failed when try to compile as shared library one.

The following error is what I got, similar to result from #499 's [comment](https://github.com/godotengine/godot-docs/issues/499#issuecomment-352264899) (thanks to hungrymonkey for taking time to test it on macOS).

```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
[ 53%] Linking Static Library ==> modules/libmodules.osx.tools.32.a
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(godot_c.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(kssl.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(s2_enc.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(s2_pkt.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(s2_srvr.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(s2_clnt.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(s2_lib.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(e_rc5.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(m_md2.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(ebcdic.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(cms_cd.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistp224.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistp521.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistputil.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistp256.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(v3_asid.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(v3_addr.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(fips_ers.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(rand_os2.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(rand_nw.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(rand_win.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(debug.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(wincerts.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(cpu.osx.tools.32.o) has no symbols
/opt/local/bin/ranlib: file: modules/libmodules.osx.tools.32.a(lookup.osx.tools.32.o) has no symbols
Ranlib Library         ==> modules/libmodules.osx.tools.32.a
ranlib: file: modules/libmodules.osx.tools.32.a(godot_c.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(kssl.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(s2_enc.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(s2_pkt.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(s2_srvr.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(s2_clnt.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(s2_lib.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(e_rc5.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(m_md2.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(ebcdic.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(cms_cd.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistp224.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistp521.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistputil.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(ecp_nistp256.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(v3_asid.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(v3_addr.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(fips_ers.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(rand_os2.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(rand_nw.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(rand_win.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(debug.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(wincerts.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(cpu.osx.tools.32.o) has no symbols
ranlib: file: modules/libmodules.osx.tools.32.a(lookup.osx.tools.32.o) has no symbols
[ 97%] Linking Program        ==> bin/godot.osx.tools.32
ld: library not found for -lsumator.osx.tools.32
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.osx.tools.32] Error 1
scons: building terminated because of errors.
```

Please advice. Thank you very much.

**Comment by vnen** [comment]
You should run the SCons command with `dev=yes` so it shows warnings and the command invocation.

**Comment by haxpor** [comment]
I executed `scons sumator_shared=yes dev=yes platform=osx` then the relevant log is at the end as follows.

```
...
[ 97%] g++ -o bin/godot.osx.tools.32 -arch i386 -framework Cocoa -framework Carbon -framework OpenGL -framework AGL -framework AudioUnit -framework CoreAudio -lz -framework IOKit -framework ForceFeedback -mmacosx-version-min=10.9 platform/osx/os_osx.osx.tools.32.o platform/osx/godot_main_osx.osx.tools.32.o platform/osx/audio_driver_osx.osx.tools.32.o platform/osx/sem_osx.osx.tools.32.o platform/osx/dir_access_osx.osx.tools.32.o platform/osx/joystick_osx.osx.tools.32.o -Lbin main/libmain.osx.tools.32.a main/tests/libtests.osx.tools.32.a modules/libmodules.osx.tools.32.a drivers/libdrivers.osx.tools.32.a editor/libeditor.osx.tools.32.a scene/libscene.osx.tools.32.a servers/libservers.osx.tools.32.a core/libcore.osx.tools.32.a modules/freetype/libfreetype_builtin.osx.tools.32.a -lpthread -lsumator.osx.tools.32
ld: library not found for -lsumator.osx.tools.32
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.osx.tools.32] Error 1
scons: building terminated because of errors.
```

Not sure that it's a problem with `-lsumator.osx.tools.32` as from the script that built file should be in `bin/` directory. Anyway, script code is the same as in document.

**Comment by hungrymonkey** [comment]
@haxpor I think the issue is in godot engine really.  godot-docs is probably not the correctly place to report it

**Comment by haxpor** [comment]
I closed this now as I moved on to 3.0, and I successfully compile as static or shared library. Thanks.
